src/pkgcheck/checks/codingstyle.py

```diff
@@ -903,3 +903,71 @@ def feed(self, pkg):
                         yield RedundantDodir(
                             cmd.group('cmd'), line=dodir.group('call'),
                             lineno=lineno - 1, pkg=pkg)
+
+
+class UnquotedVariable(results.LineResult, results.Style):
+    """Variable is unquoted, but should be quoted"""
+
+    def __init__(self, variable, **kwargs):
+        super().__init__(**kwargs)
+        self.variable = variable
+
+    @property
+    def desc(self):
+        return f"{self.lineno}: Variable {self.variable} should be quoted"
+
+
+class UnquotedVariablesCheck(Check):
+    """Scan ebuild for variables that should be quoted like D, FILESDIR, etc."""
+
+    _source = sources.EbuildParseRepoSource
+    known_results = frozenset([UnquotedVariable])
+
+    message_commands = (
+        "die", "echo", "eerror", "einfo", "elog", "eqawarn", "ewarn"
+    )
+    var_names = (
+        "D", "DISTDIR", "FILESDIR", "S", "T", "ROOT", "BROOT", "WORKDIR",
+        # EAPI=3/Prefix vars
+        "ED", "EPREFIX", "EROOT",
+        # variables for games.eclass
```

Comment:
Thankfully, we killed off `games.eclass` a little while ago, although some ebuilds still repurpose these variables. Might be with just dropping these.

Also: cc @ionenwks in case he has any further ideas for variables here.

Reply:
Nothing to add, mgorny already noted.

wrt eclasses, like mentioned on IRC, if really wanted to go there with variables it'd probably make more sense to annotate the need for quotes in eclasses themselves, e.g. `@NEED_QUOTES` `@MAY_CONTAIN_SPACES` or something like that so that docs would inform + pkgcheck check based on it. But BUILD_DIR is so common that I'd say it's a good one to have for now.

Reply:
Ah, sorry, I hadn't read the full -portage backlog. (I edited comment but was interested in your take on eclasses, but that's sorted now.) Makes sense, thanks.

Reply:
So we don't want to check eclasses just yet? This check only works on ebuilds as it is.

Reply:
Sorry, was referring to variables defined by eclasses. I think we'd like to check eclasses too, yeah, if possible.

Reply:
I'll add that in.

Reply:
Just pushed an expansion to check eclasses as well.